specs/waku/v2/waku-spam-protected-relay.md

```diff
@@ -0,0 +1,130 @@
+
+Note that the `authPath` of each peer depends on the current status of the registration tree (hence changes when new peers register). As such, it is recommended (and necessary for anonymity) that the publisher updates her `authPath` based on the latest status of the tree and attempts the proof using her updated `authPath`.
+
+
+## Routing
+
+Upon the receipt of a PubSub message, the routing peer needs to extract and parse the `proofBundle` from the `data` field.  If the `epoch` attached to the message has a non-reasonable gap (TODO: the gap should be defined) with the routing peer's current `epoch` then the message must be dropped (this is to prevent a newly registered peer spamming the system by messaging for all the past epochs). 
+Furthermore, the routing peers MUST check whether the `proofBundle` is valid and the message is not spam. If both checks are passed successfully, then the message is relayed. If `proofBundle` is invalid then the message is dropped. If spamming is detected, the publishing peer gets slashed. 
+
+### Spam Detection and Slashing
+In order to enable local spam detection and slashing, routing peers MUST record the `internalNullifier`, `shareX`, and `shareY` of any incoming message conditioned that it is not spam and has valid proof. To do so, the peer should follow the following steps. 
+1. The routing peer first verifies the `zkProof` and drops the message if not verified. 
+2. Otherwise, it checks whether a message with an identical `internalNullifier` has already been relayed. 
+   1. If such message exists and its `shareX` and `shareY` components are different from the incoming message, then slashing takes place (if the `shareX` and `shareY` fields of the previously relayed message is identical to the incoming message, then the message is a duplicate and shall be dropped).
+   2. If none found, then the message gets relayed.
+
+
+
+TODO: may shorten or delete the Spam detection and slashing process
+
+TODO: may consider [validator functions](https://github.com/libp2p/specs/tree/master/pubsub#topic-validation) or [extended validators](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md#extended-validators) for the spam detection
+
+# Security Considerations
+TODO: add discussion about the anonymity (e.g., the `StrictNoSign` policy)
+
+-------
+
+# Protobuf
+
+```protobuf
+//TODO may include the pubsub message 
+message WakuMessage {
+  optional bytes payload = 1;
+  optional fixed32 contentTopic = 2;
+  optional uint32 version = 3;
+  optional ProofBundle proofBundle = 4;
```

Comment:
I think in proto3 `optional` is omitted as in `ProofBundle` below.

Reply:
I agree, I think after the removal of `required` in proto3 the `optional` declaration is also redundant , but I wanted to be consistent with the general style of other waku specs :)

Reply:
Indeed, I borrowed the WakuMessage protobuf from its own spec, that is why `optional` is still there

Reply:
Yeah we should probably clean that up. It is worth noting that IIRC libp2p uses a mix of proto2 and proto3, and aren't always explicit about it